File: skills/trace-state-machine-port-verification/templates/cpp/trace.cpp
// Implementation of the C++ trace helper. Copy into the C++ project.
// See ../../reference/trace-contract.md.
//
// Design notes:
//  - Open TRACE_OUT with truncation once, on first use.
//  - Flush after every event so traces survive crashes / non-zero exits.
//  - A single mutex guards both the seq counter and the write, so concurrent
//    callers cannot interleave a line or reuse a seq.
//  - Do NOT emit pointers, wall-clock time, or internal symbol names.
#include "trace.h"

#include <cstdlib>
#include <fstream>
#include <mutex>

namespace trace {
namespace {

std::mutex g_mutex;
std::ofstream g_out;
bool g_init = false;
bool g_enabled = false;
long g_seq = 0;

void ensure_open() {
  if (g_init) return;
  g_init = true;
  const char* path = std::getenv("TRACE_OUT");
  if (path && *path) {
    g_out.open(path, std::ios::out | std::ios::trunc);
    g_enabled = g_out.is_open();
  }
}

std::string escape(const std::string& s) {
  std::string out;
  for (char c : s) {
    if (c == '"' || c == '\\') out += '\\';
    out += c;
  }
  return out;
}

}  // namespace

void trace_event(const std::string& event,
                 std::initializer_list<Param> params) {
  std::lock_guard<std::mutex> lock(g_mutex);
  ensure_open();
  if (!g_enabled) return;

  const char* run_id = std::getenv("RUN_ID");
  g_out << "{\"version\":1"
        << ",\"run_id\":\"" << escape(run_id ? run_id : "cpp-run") << "\""
        << ",\"seq\":" << ++g_seq
        << ",\"impl\":\"cpp\""
        << ",\"event\":\"" << escape(event) << "\""
        << ",\"params\":{";
  bool first = true;
  for (const auto& p : params) {
    if (!first) g_out << ",";
    first = false;
    g_out << "\"" << escape(p.first) << "\":\"" << escape(p.second) << "\"";
  }
  g_out << "}}\n";
  g_out.flush();
}

}  // namespace trace

// Example call sites (delete — for reference only):
//   trace::trace_event("Start", {{"session", session_id}});
//   trace::trace_event("HeaderParsed", {{"session", session_id}, {"kind", kind}});
//   trace::trace_event("ErrorRaised", {{"code", error_code}});

File: skills/trace-state-machine-port-verification/templates/rust/trace.rs
//! Minimal trace instrumentation for the Rust port.
//! Copy into the Rust project (e.g. `src/trace.rs`) and adapt.
//! See ../../reference/trace-contract.md for the format and output contract.
//!
//! Emits JSON Lines to the file named by the `TRACE_OUT` env var. When
//! `TRACE_OUT` is unset, tracing is a no-op (zero behavioral impact).
//!
//! Use the SAME abstract event names and param shapes as the C++ side. Keep
//! events semantic — never emit pointers, wall-clock time, or internal symbols.
//!
//! If a resource-release event is emitted from `Drop`, make sure an explicit
//! `close`/`release` path does not ALSO emit it — avoid double-counting.

use std::fs::File;
use std::io::Write;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Mutex, OnceLock};

struct Sink {
    file: Mutex<File>,
    run_id: String,
}

static SINK: OnceLock<Option<Sink>> = OnceLock::new();
static SEQ: AtomicU64 = AtomicU64::new(0);

fn sink() -> Option<&'static Sink> {
    SINK.get_or_init(|| {
        let path = std::env::var("TRACE_OUT").ok().filter(|p| !p.is_empty())?;
        let file = File::create(path).ok()?; // create() truncates
        let run_id = std::env::var("RUN_ID").unwrap_or_else(|_| "rust-run".into());
        Some(Sink { file: Mutex::new(file), run_id })
    })
    .as_ref()
}

fn escape(s: &str) -> String {
    s.replace('\\', "\\\\").replace('"', "\\\"")
}

/// Emit one event. `params` is a slice of (key, value) string pairs.
/// Thread-safe; flushes after every event so traces survive panics/early exits.
pub fn trace_event(event: &str, params: &[(&str, &str)]) {
    let Some(sink) = sink() else { return };
    let seq = SEQ.fetch_add(1, Ordering::SeqCst) + 1;

    let mut line = format!(
        "{{\"version\":1,\"run_id\":\"{}\",\"seq\":{},\"impl\":\"rust\",\"event\":\"{}\",\"params\":{{",
        escape(&sink.run_id),
        seq,
        escape(event),
    );
    for (i, (k, v)) in params.iter().enumerate() {
        if i > 0 {
            line.push(',');
        }
        line.push_str(&format!("\"{}\":\"{}\"", escape(k), escape(v)));
    }
    line.push_str("}}\n");

    let mut f = sink.file.lock().unwrap();
    let _ = f.write_all(line.as_bytes());
    let _ = f.flush();
}

// Example call sites (delete — for reference only):
//   trace_event("Start", &[("session", &session_id)]);
//   trace_event("HeaderParsed", &[("session", &session_id), ("kind", kind)]);
//   trace_event("ErrorRaised", &[("code", &error_code)]);
